test/cpp/microbenchmarks/bm_threadpool.cc

```diff
@@ -0,0 +1,387 @@
+#include "test/cpp/microbenchmarks/helpers.h"
+#include "test/cpp/util/test_config.h"
+
+namespace grpc {
+namespace testing {
+
+// This helper class allows a thread to block for a pre-specified number of
+// actions. BlockingCounter has an initial non-negative count on initialization.
+// Each call to DecrementCount will decrease the count by 1. When making a call
+// to Wait, if the count is greater than 0, the thread will be blocked, until
+// the count reaches 0.
+class BlockingCounter {
+ public:
+  BlockingCounter(int count) : count_(count) {}
+  void DecrementCount() {
+    std::lock_guard<std::mutex> l(mu_);
+    count_--;
+    if (count_ == 0) cv_.notify_all();
+  }
+
+  void Wait() {
+    std::unique_lock<std::mutex> l(mu_);
+    while (count_ > 0) {
+      cv_.wait(l);
+    }
+  }
+
+ private:
+  int count_;
+  std::mutex mu_;
+  std::condition_variable cv_;
+};
+
+// This is a functor/closure class for threadpool microbenchmark.
+// This functor (closure) class will add another functor into pool if the
+// number passed in (num_add) is greater than 0. Otherwise, it will decrement
+// the counter to indicate that task is finished. This functor will suicide at
+// the end, therefore, no need for caller to do clean-ups.
+class AddAnotherFunctor : public grpc_experimental_completion_queue_functor {
+ public:
+  AddAnotherFunctor(grpc_core::ThreadPool* pool, BlockingCounter* counter,
+                    int num_add)
+      : pool_(pool), counter_(counter), num_add_(num_add) {
+    functor_run = &AddAnotherFunctor::Run;
+    internal_next = this;
+    internal_success = 0;
+  }
+  // When the functor gets to run in thread pool, it will take itself as first
+  // argument and internal_success as second one.
+  static void Run(grpc_experimental_completion_queue_functor* cb, int ok) {
+    auto* callback = static_cast<AddAnotherFunctor*>(cb);
+    if (--callback->num_add_ > 0) {
+      callback->pool_->Add(new AddAnotherFunctor(
+          callback->pool_, callback->counter_, callback->num_add_));
+    } else {
+      callback->counter_->DecrementCount();
+    }
+    // Suicides.
+    delete callback;
+  }
+
+ private:
+  grpc_core::ThreadPool* pool_;
+  BlockingCounter* counter_;
+  int num_add_;
+};
+
+void ThreadPoolAddAnotherHelper(benchmark::State& state,
+                                int concurrent_functor) {
+  const int num_iterations = state.range(0);
+  const int num_threads = state.range(1);
+  // Number of adds done by each closure.
+  const int num_add = num_iterations / concurrent_functor;
+  grpc_core::ThreadPool pool(num_threads);
+  while (state.KeepRunningBatch(num_iterations)) {
+    BlockingCounter counter(concurrent_functor);
+    for (int i = 0; i < concurrent_functor; ++i) {
+      pool.Add(new AddAnotherFunctor(&pool, &counter, num_add));
+    }
+    counter.Wait();
+  }
+  state.SetItemsProcessed(state.iterations());
+}
+
+static void BM_ThreadPool1AddAnother(benchmark::State& state) {
```

Comment:
Does the BENCHMARK macro work with templates? This would be a natural usecase for them.